go-server-server/main.go

```diff
@@ -15,73 +15,146 @@ import (
-        log.Printf("trace: HTTP server Shutdown: %v", err)
-    } else {
-        log.Printf("info: Server shutdown successful!")
+    messenger <- 0
+    return
+}
+
+func monitor_certs(messenger chan<- int, wgroup *sync.WaitGroup) {
+    defer wgroup.Done()
+    client_cert_finfo, _ := os.Lstat(*sw.ClientCertFlag)
+    prev_client_cert_mtime := client_cert_finfo.ModTime()
+    log.Printf("trace: Last modified time of %s is %d", client_cert_finfo.Name(), prev_client_cert_mtime.Unix())
+
+    server_cert_finfo, _ := os.Lstat(*sw.ServerCertFlag)
+    prev_server_cert_mtime := server_cert_finfo.ModTime()
+    log.Printf("trace: Last modified time of %s is %d", server_cert_finfo.Name(), prev_server_cert_mtime.Unix())
+
+    sever_key_finfo, _ := os.Lstat(*sw.ServerKeyFlag)
+    prev_sever_key_mtime := sever_key_finfo.ModTime()
+    log.Printf("trace: Last modified time of %s is %d", sever_key_finfo.Name(), prev_sever_key_mtime.Unix())
+
+    time.Sleep(CERT_MONITOR_FREQUENCY)
+
+    for {
+        reload := false
+        client_cert_finfo, _ := os.Lstat(*sw.ClientCertFlag)
+        client_cert_mtime := client_cert_finfo.ModTime()
+        log.Printf("trace: Last modified time of %s is %d", client_cert_finfo.Name(), client_cert_mtime.Unix())
+        if client_cert_mtime != prev_client_cert_mtime {
+            log.Printf("info: Last modified time of %s changed from %d to %d", client_cert_finfo.Name(), prev_client_cert_mtime.Unix(), client_cert_mtime.Unix())
+            reload = true
+        }
+        prev_client_cert_mtime = client_cert_mtime
+
+        server_cert_finfo, _ := os.Lstat(*sw.ServerCertFlag)
+        server_cert_mtime := server_cert_finfo.ModTime()
+        log.Printf("trace: Last modified time of %s is %d", server_cert_finfo.Name(), server_cert_mtime.Unix())
+        if server_cert_mtime != prev_server_cert_mtime {
+            log.Printf("info: Last modified time of %s changed from %d to %d", server_cert_finfo.Name(), prev_server_cert_mtime.Unix(), server_cert_mtime.Unix())
+            reload = true
+        }
+        prev_server_cert_mtime = server_cert_mtime
+
+        sever_key_finfo, _ := os.Lstat(*sw.ServerKeyFlag)
+        sever_key_mtime := sever_key_finfo.ModTime()
+        log.Printf("trace: Last modified time of %s is %d", sever_key_finfo.Name(), sever_key_mtime.Unix())
+        if sever_key_mtime != prev_sever_key_mtime {
+            log.Printf("info: Last modified time of %s changed from %d to %d", sever_key_finfo.Name(), prev_sever_key_mtime.Unix(), sever_key_mtime.Unix())
+            reload = true
+        }
+        prev_sever_key_mtime = sever_key_mtime
+
+        if reload == true {
+            log.Printf("info: Certs have rolled! Reload needed!")
+            messenger <- 1
+        }
+        time.Sleep(CERT_MONITOR_FREQUENCY)
     }
-    os.Exit(0)
 }
 
 func main() {
     iniflags.Parse()
 
     sw.InitLogging()
+    var wgroup sync.WaitGroup
+    var messenger = make(chan int, 1)
```

Comment:
Can we also use unbuffered channel here?  Just a clarification

Reply:
So, just to clarify, IMO, this use-case is more aligned towards the usage of an unbuffered channel as compared to buffered. Hence clarifying on the rationale.

Reply:
If cert_monitor and signal handler both get triggered at the same time, unbuffered channel will make either of them blocked until the message is serviced.

Reply:
Isn't that we want here?

Reply:
Why would we want to block? They can always complete execution by pushing a message and end the go-routine.